Add Check for PQCPrivateKey in Decapsulator

johnpeck-us-ibm opened the pull request:
Added code to check the private key passed in and
added tests to check the code for invalid keys

Signed-off-by: John Peck <140550562+johnpeck-us-ibm@users.noreply.github.com>

jasonkatonica commented on src/main/java/com/ibm/crypto/plus/provider/MLKEMImpl.java:
```diff
@@ -129,7 +129,7 @@ public KEMSpi.DecapsulatorSpi engineNewDecapsulator(PrivateKey privateKey,
             throws InvalidAlgorithmParameterException, InvalidKeyException {
  
-        if (privateKey == null) {
+        if (privateKey == null || !(privateKey instanceof PQCPrivateKey)) {
```
I do understand the nature of this part of the update to ensure we have a PQCPrivateKey. Could we in the future, perhaps another update support private key material from OpenJDK here converting the key to our format?

jasonkatonica replied:
Yes if we have a means to do this automatically here then that would be ideal. I would assume that the conversion will also make a check very similar to this such that if it was not an appropriate key ( Such as RSA in the test below ) it would also fail given the key is not appropriate to be converted.

jasonkatonica replied:
Great in the long run this prob would be better.

jasonkatonica commented on src/test/java/ibm/jceplus/junit/base/BaseTestKEM.java:
```diff
@@ -175,6 +176,49 @@ public void testKEMSmallerSecret(String Algorithm) throws Exception {
+            KEM.Encapsulator encr = kem.newEncapsulator(pqcKeyPair.getPublic());
+            fail("testKEMKeys failed RSA public key worked.");
+        } catch (InvalidKeyException ike) {
+            // continue with test
```
Asserting the error message here and other catch statements of "unsupported key" might be more specific.

johnpeck-us-ibm replied:
See updates to messages in code. They were changes to be specific to each case.

johnpeck-us-ibm replied:
Added.

jasonkatonica reviewed (commented):
Some tests are also failing with the following unused variables:
```
Error:  /home/runner/work/OpenJCEPlus/OpenJCEPlus/src/test/java/ibm/jceplus/junit/base/BaseTestKEM.java:190:16: Unused local variable 'encr'. [UnusedLocalVariable]
Error:  /home/runner/work/OpenJCEPlus/OpenJCEPlus/src/test/java/ibm/jceplus/junit/base/BaseTestKEM.java:197:16: Unused local variable 'decr'. [UnusedLocalVariable]
Error:  /home/runner/work/OpenJCEPlus/OpenJCEPlus/src/test/java/ibm/jceplus/junit/base/BaseTestKEM.java:208:16: Unused local variable 'encr'. [UnusedLocalVariable]
Error:  /home/runner/work/OpenJCEPlus/OpenJCEPlus/src/test/java/ibm/jceplus/junit/base/BaseTestKEM.java:215:16: Unused local variable 'decr'. [UnusedLocalVariable]
```

jasonkatonica commented on src/test/java/ibm/jceplus/junit/base/BaseTestKEM.java:
```diff
@@ -175,6 +168,60 @@ public void testKEMSmallerSecret(String Algorithm) throws Exception {
+            KEM.Encapsulator encr = kem.newEncapsulator(pub);
+            if (removeWarning) {
+                System.out.println((Object) encr);
+            }
```
No need to print each `encr` here unless you have a reason?

To remove the unused variable we could replace:
```
EM.Encapsulator encr = kem.newEncapsulator(pub);
            if (removeWarning) {
                System.out.println((Object) encr);
            }
```
with:
```
kem.newEncapsulator(pub);
```
Similar question on the other spots we print the decapsulator.

johnpeck-us-ibm replied:
I fixed this by using .toString() with out printing anything.

jasonkatonica replied:
Yes this does fix the warning but we can remove all the code and just not declare `EM.Encapsulator encr` at all. In this case we wont get the warning and we can still make a call to the method under test with just `kem.newEncapsulator(pub);`

jasonkatonica commented on src/test/java/ibm/jceplus/junit/base/BaseTestKEM.java:
```diff
@@ -175,6 +169,51 @@ public void testKEMSmallerSecret(String Algorithm) throws Exception {
+
+        try {
+            KEM.Encapsulator encr = kem.newEncapsulator(pqcKeyPair.getPublic());
+            encr.toString();
```
I dont think we need to execute `.toString()` on `encr`. We also dont need encr. I think we can change these lines:
```
KEM.Encapsulator encr = kem.newEncapsulator(pqcKeyPair.getPublic());
encr.toString();
```
to just:
```
kem.newEncapsulator(pqcKeyPair.getPublic());
```
The important part here is we call the `newEncapsulator` method to raise the exception we are driving.

johnpeck-us-ibm replied:
I am doing this to get rid of a compile warning. I need to make use of encr and decr to get rid of the compiler warning. So, doing the toString makes the compiler happy. I probably code use KEM.Encapsulator_ instead of KEM.Encapsulator encr but that would only work on 25 and 26.

jasonkatonica replied:
I switched to your branch and made [the above update ](https://github.com/IBM/OpenJCEPlus/pull/1042/files#r2615314875)in various places including these lines and did not see any sort of compilation warning. The checkstyle failure from earlier was also removed. Could you post more on the compiler error? Printing shoudl not be required from the compiler.

jasonkatonica commented on src/test/java/ibm/jceplus/junit/base/BaseTestPQCKeyInterop.java:
```diff
@@ -74,6 +74,38 @@ public void testPQCKeyGenKEM_PlusToInterop() throws Exception {
+        KEM.Decapsulator decr = kemInterop.newDecapsulator(privateKey);
+        SecretKey keyD = decr.decapsulate(enc.encapsulation(), 0, 32, "AES");
+
+        assertTrue(Arrays.equals(keyE.getEncoded(), keyD.getEncoded()), "Secrets do NOT match");
```
Junit5 assert array can print what the arrays are in the case of failure. So instead of:
```
assertTrue(Arrays.equals(keyE.getEncoded(), keyD.getEncoded()), "Secrets do NOT match");
```
This might be more informative:
```
assertArrayEquals(keyE.getEncoded(), keyD.getEncoded(), "Secrets do NOT match");
```

johnpeck-us-ibm replied:
Updated All

jasonkatonica commented on src/main/java/com/ibm/crypto/plus/provider/MLKEMImpl.java:
```diff
@@ -60,14 +64,28 @@ private int getEncapsulationLength() {
+                pubKey = kf.generatePublic(publicKeySpec);
+       
+            } catch (Exception e) {
+                throw new InvalidKeyException("unsupported key", e.getCause());
```
We want e here not the cause of e.

johnpeck-us-ibm committed:
Add Check for PQCPrivateKey in Decapsulator
Added code to check the private key passed in and
added tests to check the code for invalid keys

Signed-off-by: John Peck <140550562+johnpeck-us-ibm@users.noreply.github.com>

jasonkatonica reviewed (approved):
LGTM

KostasTsiounis commented on src/test/java/ibm/jceplus/junit/base/BaseTestKEM.java:
```diff
@@ -175,6 +169,47 @@ public void testKEMSmallerSecret(String Algorithm) throws Exception {
+            fail("testKEMKeys failed - NULL Private key did not cause an Invalid Key Excepton.");
+        } catch (InvalidKeyException ike) {
+            assertTrue(ike.getMessage().equals("unsupported key"));
+        }
```
Would it make sense to separate this into another test method?

johnpeck-us-ibm replied:
It is doing testing of the keys and we really expect the same Exception. So, I don't thing it needs to be broken up.

johnpeck-us-ibm commented on src/main/java/com/ibm/crypto/plus/provider/MLKEMImpl.java:
```diff
@@ -129,14 +147,27 @@ public KEMSpi.DecapsulatorSpi engineNewDecapsulator(PrivateKey privateKey,
+        PrivateKey privKey = privateKey;
+
+        if (privKey == null) {
             throw new InvalidKeyException("unsupported key");
```
Updated

johnpeck-us-ibm commented on src/main/java/com/ibm/crypto/plus/provider/MLKEMImpl.java:
```diff
@@ -60,14 +64,28 @@ private int getEncapsulationLength() {
+        
+        PublicKey pubKey = publicKey;
+        if (pubKey == null) {
             throw new InvalidKeyException("unsupported key");
```
Updated

johnpeck-us-ibm commented on src/main/java/com/ibm/crypto/plus/provider/MLKEMImpl.java:
```diff
@@ -129,14 +147,27 @@ public KEMSpi.DecapsulatorSpi engineNewDecapsulator(PrivateKey privateKey,
+            // Try and convert this key to a usage PQCPrivateKey
+            try {
+                KeyFactory kf = KeyFactory.getInstance(this.alg, this.provider.getName());      
+                PKCS8EncodedKeySpec privateKeySpec = new PKCS8EncodedKeySpec(privateKey.getEncoded());
```
My understanding of clone is that the result of clone is another separate object with the same value as the original. This is done so that changes can be done to the cloned instance with out affecting the original. So, there is no way to clear the encoding from PKCS8EncodedKeySpec.

KostasTsiounis replied:
The call to `privateKey.getEncoded()` creates an encoding of the private key that is saved in memory. The constructor of `PKCS8EncodedKeySpec` though does a clone on that (doesn't use the original) and the original byte array is dereferenced and not used any more but the encoding is not zeroed out. The only way to do that is with something similar to what I suggested above.

johnpeck-us-ibm replied:
I get what you are saying. I will fix it like you suggested.

KostasTsiounis commented on src/test/java/ibm/jceplus/junit/base/BaseTestPQCKeyInterop.java:
```diff
@@ -74,6 +75,38 @@ public void testPQCKeyGenKEM_PlusToInterop() throws Exception {
+    public void testPQCKeyGenKEMAutoKeyConvertion() throws Exception {
+        String pqcAlgorithm = "ML-KEM-512";
+
+        if (getProviderName().equals("OpenJCEPlusFIPS") || 
```
We could substitute this with:
```
assumeFalse("OpenJCEPlusFIPS".equals(getProviderName()));
assumeFalse(Utils.PROVIDER_BC.equals(getProviderName()));
```

johnpeck-us-ibm replied:
Updated

johnpeck-us-ibm commented on src/test/java/ibm/jceplus/junit/base/BaseTestKEM.java:
```diff
@@ -175,6 +169,47 @@ public void testKEMSmallerSecret(String Algorithm) throws Exception {
+  
+        try {
+            kem.newDecapsulator(priv);
+            fail("testKEMKeys failed - NULL Private key did not cause an Invalid Key Excepton.");
```
Fixed

johnpeck-us-ibm commented on src/test/java/ibm/jceplus/junit/base/BaseTestKEM.java:
```diff
@@ -175,6 +169,47 @@ public void testKEMSmallerSecret(String Algorithm) throws Exception {
+  
+        try {
+            kem.newDecapsulator(pqcKeyPair.getPrivate());
+            fail("testKEMKeys failed - RSA Private key did not cause an InvalidKeyExcepton.");
```
Hope I got it right this time :(

johnpeck-us-ibm commented on src/test/java/ibm/jceplus/junit/base/BaseTestKEM.java:
```diff
@@ -175,6 +169,47 @@ public void testKEMSmallerSecret(String Algorithm) throws Exception {
+
+        try {
+            kem.newEncapsulator(pub);
+            fail("testKEMKeys failed - NULL Public key did not cause an InvalidKeyExcepton.");
```
Hope I got it right this time :(

KostasTsiounis reviewed (approved):
LGTM
